src/nvim/menu.c

```diff
@@ -1301,11 +1301,15 @@ void ex_emenu(exarg_T *eap)
 
   /* Found the menu, so execute.
    * Use the Insert mode entry when returning to Insert mode. */
-  if (restart_edit
+  if (
+      ((State & INSERT) || restart_edit)
       && !current_SID
       ) {
     mode = (char_u *)"Insert";
     idx = MENU_INDEX_INSERT;
+  } else if (get_real_state() & VISUAL) {
```

Comment:
That's what I thought at first, but it seems that block is all about guessing the relevant visual-mode state and recreating it temporarily before running the command. If we're already in 'real' visual mode, doing that stuff leaves us with an incorrect selection (it seemed to always select the entire buffer in my tests)